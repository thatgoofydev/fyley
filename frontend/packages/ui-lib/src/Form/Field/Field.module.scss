@import "src/base-styling/colors";

.formControl {
  font-size: 12px;
  display: flex;
  width: 460px;

  label {
    display: block;
    position: absolute;
    align-self: center;
    padding-left: 14px;

    font-weight: bold;
    text-transform: uppercase;
    color: #9096A9;
    user-select: none;
    cursor: text;
  }

  input {
    width: 100%;
    padding: 14px 14px 14px 114px;
    border: 1px solid $grey;
    border-radius: 2px;

    box-shadow: none !important;

    &:focus {
      outline: none !important;
      border-color: $mid-grey;
    }
  }
}

.error {
  label {
    color: $red;
  }

  input {
    border-color: $red;
  }
}

.errorText {
  margin: 4px 0 0 14px;
  color: $red;
  font-size: 12px;
}