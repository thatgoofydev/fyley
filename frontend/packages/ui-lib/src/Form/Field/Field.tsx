import React, { FunctionComponent } from "react";
import { useFieldState } from "../useFieldState";
import styles from "./Field.module.scss";
import classNames from "classnames";
import { FormControl } from "../FormControl/FormControl";

interface IFieldProps {
  name: string;
  label: string;
  placeholder?: string;
}

export const Field: FunctionComponent<IFieldProps> = ({ name, label, placeholder }) => {
  const { inContext, onChange, onFocus, onBlur, value, error, focused, touched } = useFieldState(
    name
  );

  if (!inContext) {
    console.error("Field should be used inside a 'Form' component");
    return <></>;
  }

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => onChange(event.target.value);

  const showError = error && touched;
  const classes = classNames({ [styles.error]: showError });
  return (
    <>
      <FormControl className={classes}>
        <div className={styles.formControl}>
          <label htmlFor={name}>{label}</label>
          <input
            id={name}
            type="text"
            name={name}
            value={value || ""}
            placeholder={placeholder}
            onChange={handleChange}
            onFocus={onFocus}
            onBlur={onBlur}
            autoComplete="off"
          />
        </div>
        {showError && <p className={styles.errorText}>{error}</p>}
      </FormControl>
    </>
  );
};
