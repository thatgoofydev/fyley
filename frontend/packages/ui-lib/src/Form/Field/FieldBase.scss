@import "src/base-styling/colors";

@mixin labelBase() {
  display: block;
  color: $primary-text;
  margin-bottom: 4px;
}

@mixin inputBase() {
  display: block;
  background-color: #FAFBFC;
  border: 1px solid #E1E4E8;
  border-radius: 4px;
  width: 100%;

  padding: 7px 14px;
  color: $primary-text;
}

@mixin errorTextBase() {
  margin: 4px 0 0 14px;
  color: $red;
  font-size: 12px;
}