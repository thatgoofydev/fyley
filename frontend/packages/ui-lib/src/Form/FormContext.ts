import { createContext } from "react";
import { IFormContext } from "./types";

const noop = () => {};

export const FormContext = createContext<IFormContext>({
  state: {
    values: {},
    errors: {},
    focused: {},
    touched: {}
  },
  getFieldState: () => ({
    value: {},
    error: {},
    focused: false,
    touched: false
  }),
  setFieldValue: noop,
  onFieldFocus: noop,
  onFieldBlur: noop
});
