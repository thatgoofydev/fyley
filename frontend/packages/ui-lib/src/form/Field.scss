@import "src/base/vars";

.field {
  display: flex;
  flex-direction: column;
  width: 300px;

  label {
    padding: 2px 0;
  }

  input {
    display: block;
    border: none;
    outline: none;

    padding: $field-padding;
    border: $field-border;
    border-radius: $field-border-radius;
    background-color: $field-bg;
  }

  .error-spacing {
    min-height: 20px;
  }

  .error-message {
    display: block;
    color: $field-error-color;
    font-size: $field-error-text-size;
    line-height: $field-error-text-size;
  }

  &.focused {
    input {
      border: $field-border--focus;
      -webkit-box-shadow: $field-focus-shadow;
      -moz-box-shadow: $field-focus-shadow;
      box-shadow: $field-focus-shadow;
    }
  }

  &.error {
    input {
      border: $field-border--error;
      -webkit-box-shadow: $field-error-shadow;
      -moz-box-shadow: $field-error-shadow;
      box-shadow: $field-error-shadow;
    }
  }
}