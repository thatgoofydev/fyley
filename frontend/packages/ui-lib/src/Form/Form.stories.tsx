import React from "react";
import { Meta, Story } from "@storybook/react";
import { Form } from "./Form";
import { Field } from "../Field";
import { FormValues } from "./types";

export default {
  title: "Form/Example",
  component: Form
} as Meta;

export const FormExample: Story = () => {
  const onSubmit = (values: FormValues) => {
    console.log(values);
  };

  const onValidate = (values: FormValues): FormValues => {
    const errors: FormValues = {};

    if (values.name !== "jeroen") {
      errors.name = "wrong name";
    }

    return errors;
  };

  return (
    <Form
      onSubmit={onSubmit}
      onValidate={onValidate}
      initialValues={{
        name: "jer",
        birthdate: "1998-09-01"
      }}
    >
      <Field label="Full name" name="name" type="text" />
      <Field label="Birthdate" name="birthdate" type="date" />

      <button type="submit">Submit</button>
    </Form>
  );
};

// export const FormExampleTwo: Story = () => {
//     const onSubmit = (values: FormValues) => {
//         console.log(values);
//     };
//
//     const onValidate = (values: FormValues): FormValues => {
//         const errors: FormValues = {};
//
//         if (values.name !== "jeroen") {
//             errors.name = "wrong name";
//         }
//
//         errors.test = "a test";
//
//         return errors;
//     };
//
//     return (
//         <Form
//             onSubmit={onSubmit}
//             onValidate={onValidate}
//             initialValues={{
//                 name: "jer",
//                 birthdate: "1998-09-01"
//             }}
//         >
//             {({ errors, getField }) => (
//                 <>
//                     <Field label="Full name" name="name" type="text" />
//                     <Field name="Birthdate" type="date" />
//
//                     <button type="submit">Submit</button>
//
//                     {JSON.stringify(getField("name"))}
//
//                     {errors.test && <p>A test error occured</p>}
//                 </>
//             )}
//         </Form>
//     );
// };
