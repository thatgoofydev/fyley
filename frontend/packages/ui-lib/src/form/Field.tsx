import React, { FunctionComponent, useContext } from "react";
import { FormContext } from "./Form";
import "./Field.scss";
import classNames from "classnames";

interface FieldProps {
  name: string;
  type: "text" | "date";
  label?: string;
}

export const Field: FunctionComponent<FieldProps> = ({
  name,
  type,
  label = name
}) => {
  const context = useContext(FormContext);
  if (!context) {
    return <h1>Field can only be used inside a &lt;Form&gt; component</h1>;
  }

  const { value, error, focused, touched } = context.getField(name);

  const handleChange = (event: React.ChangeEvent<HTMLInputElement>) => {
    context.setValue(name, event.target.value);
  };

  const handleFocus = () => {
    context.handleFocus(name);
  };

  const handleBlur = () => {
    context.handleBlur(name);
  };

  const errorState = touched && error;

  const classes = classNames("field", { error: errorState, focused });
  return (
    <>
      <div className={classes}>
        <label>{label}</label>
        <input
          type={type}
          name={name}
          value={value || ""}
          onChange={handleChange}
          onFocus={handleFocus}
          onBlur={handleBlur}
          autoComplete="off"
        />
        <div className="error-spacing">
          {errorState && <span className="error-message">{error}</span>}
        </div>
      </div>
    </>
  );
};
