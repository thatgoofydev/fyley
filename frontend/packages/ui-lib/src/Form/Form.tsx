import React, { FunctionComponent, useState } from "react";
import { FormValues, IFieldState, IFormContext, IFormState } from "./types";
import { FormContext } from "./FormContext";

interface IFormProps {
  onSubmit: (values: FormValues) => void;
  onValidate: (values: FormValues) => FormValues;
  initialValues?: FormValues;
}

export const Form: FunctionComponent<IFormProps> = ({
  onSubmit,
  onValidate,
  initialValues,
  children
}) => {
  const [state, setState] = useState<IFormState>({
    values: initialValues || {},
    errors: {},
    focused: {},
    touched: {}
  });

  const getFieldState = (name: string): IFieldState => {
    if (state.touched[name] == undefined) {
      // Key needs to be set to allow onSubmit to set them all to true.
      state.touched[name] = false;
    }

    return {
      value: state.values[name],
      error: state.errors[name],
      focused: state.focused[name],
      touched: state.touched[name]
    };
  };

  const setFieldValue = (name: string, value: any) => {
    const newState = {
      ...state,
      values: {
        ...state.values,
        [name]: value
      }
    };

    newState.errors = onValidate(newState.values);
    setState(newState);
  };

  const onFieldFocus = (name: string) => {
    setState(prevState => ({
      ...prevState,
      focused: {
        ...prevState.focused,
        [name]: true
      }
    }));
  };

  const onFieldBlur = (name: string) => {
    setState(prevState => ({
      ...prevState,
      focused: {
        ...prevState.focused,
        [name]: false
      },
      touched: {
        ...prevState.touched,
        [name]: true
      }
    }));
  };

  const handleSubmit = (e: React.FormEvent<HTMLFormElement>) => {
    e.preventDefault();

    const newTouched = state.touched;
    Object.keys(newTouched).forEach(key => (newTouched[key] = true));

    const newState = {
      ...state,
      touched: newTouched,
      errors: onValidate(state.values)
    };

    setState(newState);

    console.log(newState.errors);
    if (!newState.errors || Object.keys(newState.errors).length === 0) {
      onSubmit(newState.values);
    }
  };

  const context: IFormContext = {
    state,
    getFieldState,
    setFieldValue,
    onFieldFocus,
    onFieldBlur
  };

  return (
    <FormContext.Provider value={context}>
      <form onSubmit={handleSubmit} autoComplete="off">
        {children}
      </form>
    </FormContext.Provider>
  );
};
