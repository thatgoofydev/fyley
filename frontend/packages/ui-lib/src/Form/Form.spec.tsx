import React from "react";
import { cleanup, render } from "@testing-library/react";
import { Form } from "./Form";
import { Field } from "./Field";

const noop = () => {};
const noValidate = () => ({});

afterEach(cleanup);

describe("Form", function () {
  it("should render", function () {
    const { container } = render(
      <Form onSubmit={noop} onValidate={noValidate}>
        <Field name="name" label="Name" />
      </Form>
    );

    expect(container).toMatchSnapshot();
  });

  it("should render with initial values", function () {
    const { container } = render(
      <Form onSubmit={noop} onValidate={noValidate} initialValues={{ name: "john" }}>
        <Field name="name" label="Name" />
      </Form>
    );

    expect(container).toMatchSnapshot();
  });
});
