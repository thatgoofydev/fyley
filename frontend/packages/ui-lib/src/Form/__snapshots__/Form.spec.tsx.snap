// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Form should render 1`] = `
<div>
  <form
    autocomplete="off"
  >
    <div
      class="formControlWrapper"
    >
      <div
        class="formControl"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          autocomplete="off"
          id="name"
          name="name"
          type="text"
          value=""
        />
      </div>
    </div>
  </form>
</div>
`;

exports[`Form should render with initial values 1`] = `
<div>
  <form
    autocomplete="off"
  >
    <div
      class="formControlWrapper"
    >
      <div
        class="formControl"
      >
        <label
          for="name"
        >
          Name
        </label>
        <input
          autocomplete="off"
          id="name"
          name="name"
          type="text"
          value="john"
        />
      </div>
    </div>
  </form>
</div>
`;
